### PR TITLE
screen: Revert the DPI per screen changes.

### PR DESCRIPTION
**Pull request by Elv13**
The changes to how DPI is calculated breaks existing config and thus
have to be reverted to a saner bahavior. If a theme or code in rc.lua
specify a size, it is in pixel. If the DPI is calculated differently,
then the assumption that the size in pixel corresponds to the user
settings (aka, 96dpi or whatever is in ~/.Xresources) is to be honored
was previously true and is now false. This is a behavior change and
cannot go into a stable branch release.

As suggested by Uli, this introduces a version check to change the
behavior when Awesome 5 is released. This assumes all other changes
to make Awesome auto-magically apply the DPI to all elements are
merged until then (I seriously doubt this will happen).

Compared to the patch draft proposed by @psychon, this patch invert the `get_dpi` logic to use `beautiful.xresources` as the primary source again, restoring the previous DPI choice. 

Ref: #2091

### CODE REVIEW

**Comment by Elv13** [comment]
Note that in itself this also breaks the behavior since there was an "hardcoded" DPI per screen system before the "automatic" one. So I guess the final patch will have to be something in between fully reverting the recent DPI per screen work or adding more "if" to make sure it doesn't try to invent its own DPI.

File: lib/wibox/widget/textbox.lua [hunk 1]
```diff
@@ -107,6 +108,8 @@ end
 -- @treturn number The preferred width.
 -- @treturn number The preferred height.
 function textbox:get_preferred_size_at_dpi(dpi)
+    assert(dpi, "No DPI provided")
```
Comment (psychon) [inline lib/wibox/widget/textbox.lua]: Huh? Why these new asserts? Does this just provide better error messages than what happened before? If so, should this assert perhaps be moved to `setup_dpi`?

**Comment by psychon** [comment]
How about we just start an v4 branch (any suggestions for good names?), revert the commits in question there and are done? That seems saner than trying to add lots of version checks...
Backward-compatible development should target the v4 branch, breaking changes go to master and from time to time v4 is merged into master.

**Comment by Elv13** [comment]
I am against that because it failed twice already:

* 3.4-3.5 took 3 years, 3.5 to 4 took 4 years. I don't see any clue that
5.0 would make the cut faster. Even if it did, it would be too soon to
break the API, the userbase would revolt, again.
* During the dual branch time (3.4.x/3.5.x), no major contributors used the
stable branch, letting it decay and bugs stay unfixed. Nobody backported
[most] compatible changes during the 3.5 era (and worst, during 3.4, it had
a separate dev base and diverged). Worst still, the changes that were
backported ended up mostly untested before releases were made because
nobody used the "git" HEAD of the stable branch. We don't have the manpower
to maintain branches. You already spend a lot of time on Awesome, Daniel
and me are very active for some time, but only once in a while, not
continuously.
* We lost 80% of our users due to, among other things, the lack of
perceived progress and the big+bad API breaks once it happened. Letting the
branch diverge will just recreate that exact same situation again. Keeping
the same branch with me complaining about any changes prevents the slippery
slope from starting again. I sucks for me too, I like changes. However I
prefer Awesome to thrive by leveraging stability and keeping the code
developed by our users compatible for an extended time frame. Yes, it does
induce minor regressions from time to time, but given we are able to
release more often than 2-3 time per decade, it's a cost I consider
acceptable.
* We are still stuck getting questions about 3.4.x (8.4 years old) due to
the inability to release often. If we branch v5 and it takes 5 years to
release, then we will have to support 4.x into the 2030's. Having a clean
"release from master" + deprecation system forces us to keep rolling. Even
if it adds many `--TODO v5` and `gears.debug.deprecate` in the code.
That's, again, a cost I consider acceptable compared to forking.
* There is often very cool stuff being developed, it's disrespectful not to
let our users have them within 4 months or so. Having a single branch
enforces that new compatible features get released.
* I am profoundly skeptical of the capacity to fully support multi-DPI.
This kind of bug is the tip of the iceberg. Any changes will only make
things worst. X11 doesn't support it for apps and it can be done using the
config already. It is still unclear to me why it could not be developed
using some kind of smart zooming at the widget scale. If that happens to be
possible, then it should be done first. And if it's done, then maybe that's
enough. Letting the little minority of users who really need this implement
it in their config themselves using the tools we provide instead of making
a gigantic number of changes affecting everybody and breaking the
API/behavior. It will turn out like a dog chasing its tail or the video
game "uncanny valley" syndrome. IMO, while having per screen DPI is cool,
it's not worth all the troubles. Having the global DPI + font scaling is a
good compromise because it keeps Awesome usable for HiDPI laptop users and
moves the burden away from us.

As for the patch here, I will do it another way. As I mentioned in the
comment when I saw luacheck failed, it indeed could not work this way. I
think I will add an `awful.screen.set_multi_dpi_enabled(bool)` (default:
false) and move the remaining part of `beautiful.xresources.get_dp(s)` into
`awful/screen.lua` to avoid the dependency loop.

**Comment by Elv13** [comment]
I improved the issues in this new version

File: lib/wibox/widget/textbox.lua (hunk 1 above)
Comment (psychon) [reply lib/wibox/widget/textbox.lua]: Ping

File: lib/awful/screen.lua [hunk 2]
```diff
@@ -537,16 +590,12 @@ function screen.object.get_dpi(s)
...
-    -- We have no outputs, so guess based on the size of the root window.
-    if root and not fallback_dpi then
-        local _, h = root.size()
-        local _, hmm = root.size_mm()
-        fallback_dpi = hmm ~= 0 and h * mm_per_inch / hmm
-    end
-    return fallback_dpi or 96
+    s.data.dpi_cache = get_fallback()
+    return s.data.dpi_cache
 end
```
Comment (psychon) [inline lib/awful/screen.lua]: :+1: for the second commit

File: lib/wibox/widget/textbox.lua (hunk 1 above)
Comment (Elv13) [reply lib/wibox/widget/textbox.lua]: done

**Comment by actionless** [comment]
sorry, i've noticed the notification only now, will have a look now

**Comment by blueyed** [comment]
@Elv13 
How do you feel about covering this with tests? :)

**Comment by Elv13** [comment]
> How do you feel about covering this with tests? :)

As stated in the related issues, one of my big issue with the DPI project is how hard it is to properly test. At some point we will need a new test suit that compare each widgets at different DPI to see if the images match (as @actionless proposes). We will also have to port every single unit of measurement currently mapping to pixels (such as wibox/client/widget geometry, margins, paddings, spacings) to an abstract units. Then implement either the proposed auto-widget-scaling or the less maintainable per-widget adjustments. Then (and only then) can we start testing DPI seriously. For now testing it would only help notice how broken it is. Once all the testing is done, then maybe this could be enabled by default.

This change is a compromise. I revert the DPI per screen changes because they were (currently) untestable and started introducing regressions as soon as they got merged. On the other hand, if an user want to test the per-screen DPI support, it can still be enabled. But then, that user has to deal with the pixel vs. dpi issues him/herself. I want to state it clearly, I am *not* against DPI scaling. But I am against user visible regressions and see this as a prime example of unstable changes with little benefits.

**Comment by actionless** [comment]
@blueyed mb if remove fake ~/.Xresources from the testsuite `get_fallback()` function and related code branches will became covered?

**Comment by blueyed** [comment]
@actionless 
Yes, https://github.com/awesomeWM/awesome/pull/2121.

@Elv13 
I agree with your comment, but think that having it covered already a bit will help with developing this further.  #2121 hopefully helps here.

**Pull request merged**